### ISSUE TITLE
Table driven scenarios do not follow scenario order and are always executed last 

### ISSUE THREAD

**Issue by hbenaicha** [getgauge/gauge#1960]
**Describe the bug**
Given a Spec that has 2 scenario and one table driven scenario, the table driven scenario will trigger last even if the order is scenario then table scenario then scenario 

**To Reproduce**
Steps (or project) to reproduce the behavior:
![image](https://user-images.githubusercontent.com/49121919/114854167-1d6c3800-9e17-11eb-827c-d36d9a91b298.png)

**Implementation:**
![image](https://user-images.githubusercontent.com/49121919/114854235-307f0800-9e17-11eb-860a-00d0928a5c19.png)

You would expect that execution is as follow :

1. run scenario 1
2. run table spec
3. run scenario 2

However when you execute, scenario table is executed last, not 2nd

![image](https://user-images.githubusercontent.com/49121919/114854390-59070200-9e17-11eb-8693-2bbb499ef7e6.png)

Gauge version used is 1.1.7

**Comment by sriv** [comment]
Gauge does not promise the order of scenario execution in a spec file. You can control the order of spec execution within a suite by using the `--sort` flag, but within a spec file gauge does not explicitly give a sort order.

**Comment by hbenaicha** [comment]
Does that mean that even for a file without any scenario table, the order of execution of the scenario is not known / defined ?
That feels like a big problem that forces us to creat a scenario per fil if we want to have a sequence of execution in a specific order

**Comment by sergueis** [comment]
> Gauge does not promise the order of scenario execution in a spec file. You can control the order of spec execution within a suite by using the `--sort` flag, but within a spec file gauge does not explicitly give a sort order.

Then what about this? 
"The scenarios should execute in the order they are specified when running in serial mode." - it's from Gauge FAQ ( https://docs.gauge.org/faqs.html?os=linux&language=python&ide=vscode#what-is-the-order-of-execution-of-scenarios-in-a-specification )

Gauge took a lot of good features from Cucumber, but didn't take sequential execution order of scenarios which is a must have.

**Comment by sriv** [comment]
@hbenaicha wrote:

> Does that mean that even for a file without any scenario table, the order of execution of the scenario is not known / defined ?

Yes. And here's why:

- the scenarios being executed can change based on various parameters - the tags being applied, the table variables being used (or not), the explicit order being specified as part of `gauge run`
- gauge recommends that the scenarios be independent. This is to avoid temporal coupling between scenarios which can be a pain to discover
- note that if scenarios were dependent, and the 3rd scenario in the dependency failed, it would be pointless to run `gauge run --failed` because it will never get the dependencies included in the run.

Note that you can force gauge to follow an order by specifying the scenarios explicitly (ex: `gauge run specs/example.spec:10 specs/example.spec:5` would run scenario in line 10 before line 5). But by default  `gauge run` does not make any promises in this regard.

@sergueis wrote:

> Then what about this?

Exactly what you quoted. Gauge will honour the order you specify as arguments to `gauge run`. See my example above. 

If you passes an entire spec file or a directory, gauge will choose to decide the order and while you may see a sequence anecdotally, gauge does not guarantee that. i.e. you can use it at your own risk, the implementation is subject to change.

> Gauge took a lot of good features from Cucumber

Incorrect. Gauge took inspiration from [Twist](https://en.wikipedia.org/wiki/Twist_(software)), i.e. both gauge and twist are built by the same team. Both Twist and Cucumber started around the same time as independent initiatives. 

>  but didn't take sequential execution order of scenarios which is a must have.

Gauge does not believe in this. The reasons are stated above.

**Comment by sergueis** [comment]
> Note that you can force gauge to follow an order by specifying the scenarios explicitly (ex: gauge run specs/example.spec:10 specs/example.spec:5 would run scenario in line 10 before line 5). But by default gauge run does not make any promises in this regard.

No, it doesn't work with the example above - scenario with table is still executed last.

**Comment by sergueis** [comment]
@sriv then how do I write a test when I need to do some preparation steps (in my case it would write a separate scenario), then insert 20 records in a DB table (2nd scenario with table), and then do some processing/checking on these records (3rd scenario)? Again, explicit scenario order doesn't work in this case (ex: gauge run specs/example.spec:10 specs/example.spec:5).

**Comment by zabil** [comment]
> Then how do I write a test when I need to do some preparation steps (in my case it would be a 1st separate scenario), then insert 20 records in a DB table (2nd scenario with table), and then do some processing/checking on these records (3rd scenario)? 

In short, it's not possible. 

In Gauge, scenarios must not depend on each other. This is because 
* By design, scenarios are independent units of execution.
* They can be run in parallel.

**Comment by sergueis** [comment]
@zabil It's a pity that modern Gauge can't handle this very common use case that is easily handled by 13+ years old Cucumber.

Moreover:
1) If scenarios are independent units, why are they needed at all? One spec-file - one unit. And dir hierarchy for grouping.
2) It's totally up to me - whether to run particular spec-file in parallel or not, to run particular tags or not, to run selected scenarios or not. They are all not mandatory requirements.

**Comment by zabil** [comment]
> Gauge can't handle this very common use case that is easily handled by 13+ years old Cucumber.

I've noticed Cucumber re-appearing in comments despite an assertion earlier that Gauge and Cucumber are two different tools, that evolved very differently. So appreciate if we keep Cucumber out of the comments. 

> It's totally up to me - whether to run particular spec-file in parallel or not, to run particular tags or not, to run selected scenarios or not. They are all not mandatory requirements.

Gauge is an opinionated testing tool. To Gauge "It's more important to have a testable system than test cases". From our experience we've seen in the past that having contraints like this forces better design of test cases and the system under tests.

For that reason ordered execution of scenarios will not be supported. Hope that makes sense. I am closing this issue for now.

**Comment by sriv** [comment]
@sergueis - the use case is common, just that Gauge does not expect them to be in different scenarios as you have described.

Perhaps your notion of a Scenario does not match [Gauge's](https://docs.gauge.org/writing-specifications.html?os=linux&language=python&ide=vscode#scenario).

Depending on whether you choose to make the DB inserts visible or transparent to the end users you can use [Context](https://docs.gauge.org/writing-specifications.html?os=linux&language=python&ide=vscode#contexts)/[Teardown ](https://docs.gauge.org/writing-specifications.html?os=linux&language=python&ide=vscode#tear-down-steps)steps in the specification file, or [execution hooks ](https://docs.gauge.org/writing-specifications.html?os=linux&language=python&ide=vscode#execution-hooks )in the implementation.

**Comment by sriv** [comment]
@zabil - I propose that we leave this issue open, because it seems that Gauge does not honour the explicit order of scenarios passed to the `gauge run` command when there is a table driven scenario. 

Also worth noting that Table Driven Scenarios is an experimental feature.

**Comment by sergueis** [comment]
@zabil Regarding Cucmber, why can't I compare Gauge with any other testing framework, regardless of their evolution/goals/contributors/age/etc?

If ordered execution of scenarios is not supported, please change this FAQ section "The scenarios should execute in the order they are specified when running in serial mode." to something like "Gauge can execute scenarios in any order" - just to not mislead users.

@sriv Thank you, I know about context/teardown but they are implicitly executed before/after each scenario which is not acceptable in our case (we have more that 3 scenarios in 1 spec-file and they are supposed to be executed sequentially). We have to get rid of table-driven scenario as a workaround.

**Comment by hbenaicha** [comment]
@sriv thanks for putting this as a bug, I agree and understand what you said below, but these are still cases where users wants to run in //, wants to run only failed cases etc...the general use case and the expectation is that Gauge would be following the order of execution of scenario / table driven one.

From that perspective it is a bug, so appreciate your support with this

**Comment by hbenaicha** [comment]
note that this part though 
> Note that you can force gauge to follow an order by specifying the scenarios explicitly (ex: gauge run specs/example.spec:10
>specs/example.spec:5 would run scenario in line 10 before line 5). But by default gauge run does not make any promises in this
>regard.

was tested in the above example and didn't seem to work for the table scenario

**Comment by kga676** [comment]
Got same issue and spent 2 days for debugging, I used DataStore Hasmap
1. generate value
2. Use in step
And I got issue when with table-driven spec step 2 has not able to find a value

Thanks for thiss issue, It saved me from madness

**Comment by sergueis** [comment]
> Got same issue and spent 2 days for debugging, I used DataStore Hasmap
> 
> 1. generate value
> 2. Use in step
>    And I got issue when with table-driven spec step 2 has not able to find a value
> 
> Thanks for thiss issue, It saved me from madness

Unfortunately, one of the gauge authors has strong (and wrong) vision that scenaros in one spec-file should be totally independent, so... don't expect this bug to be fixed soon :)

**Comment by sergueis** [comment]
Don't make them dependent, just execute in order they are defined. It's
that simple.

**Comment by bpierce01** [comment]
Not sure if this issue is still being tracked, but I would also like to see scenarios execute in the order they are defined inside the spec file. 

Like how @kga676 points out, scenarios are executed in order when there is no table driven scenario. But when there is a table driven scenario, then the execution order changes. It is inconsistent implementation. I think it's very reasonable to enforce that inside a spec file, the scenarios are always executed in the order defined.
